awx/main/models/events.py

```diff
@@ -356,6 +356,14 @@ def _update_from_event_data(self):
                     job_id=self.job_id, uuid__in=failed
                 ).update(failed=True)
```

Comment:
Why not cram everything else from line 340 up to here into that `handle_success_and_failure_notifications` task, and remove that lag too? The issue really sounded like it scaled with the number of hosts, so it's extremely likely that the main cause of the race condition on the job host summaries, as opposed to the notifications.

Reply:
I'm not sure I want to background the job host summaries into a task.  Otherwise there's potentially a scenario where you *have* a `playbook_on_stats` event, but the API for summaries doesn't have data for a few seconds.  That seems weird to me, and I worry it could introduce regressions or side effects _somewhere_.

Reply:
This isn't in a transaction now, is it? If it's not in a transaction, I don't see how it makes a difference in a separate task vs. same task regarding the data correspondence you're talking about.

If it works, it works. If the integration test shows this fixes the issue, then you should move ahead with it.

I just couldn't draw the line between the issue and this, because my assumption was the the creation of job host summaries would be the driving `O(N)` cost for the number of hosts.

Reply:
Yea, maybe so.  I do agree with this, though:

> If it works, it works. If the integration test shows this fixes the issue, then you should move ahead with it.

I get a little nervous changing this code *too* much because it's so critical :).

Reply:
> Also, even if there is a higher cost here that consumes a callback worker for a period of a few seconds, it is just once per playbook run (at the end).

But it scales with the number of hosts, and not the number of tasks. Doing a simple thing with a large number of hosts is a common use case.

Delaying the playbook_on_stats also delays the `events_processing_finished` status of the job, and getting the EOF before the stats event feels like something that could confuse the UI.

But I agree with your overall sentiment - because while _some_ of this is new, the host summaries stuff isn't new, and I believe that's the bottleneck. If the new stuff was a bottleneck that could introduce regressions, but I don't see that being the case.